### ISSUE TITLE
Per request scopes in e.g. Autofac will eat resources

### ISSUE THREAD

**Issue by danielwertheim** [jbogard/MediatR#71]
Sample here: https://gist.github.com/danielwertheim/887868c17ab8cd67be44

Kind of followed your sample code... Probably have missed something. And yes, I know I could inject `Func<SomeDependency>` instead and have the handler resolve and dispose....

But how do I acomplish per-routed message scoped handlers? If not possible with MediatR, perhaps state that clearly in the sample?

**Comment by khellang** [comment]
See http://binarysculpting.com/2012/08/19/using-idisposable-with-autofac/. You need to make sure you resolve from a lifetime scope, and that the scope is disposed. Otherwise, Autofac will hold on to your transient instances and you'll have a leak.

**Comment by danielwertheim** [comment]
I know. And that's why the sample is misleading as it will consume new resources for each message routed as the gist will show.

So **the Q is how to achieve this with MediatR** in e.g. something like a long running windows service?

Can it? As you would need to get hooked into before and after the func for resolving the handler....

Just raising my voice as this easily could lead to people getting more and more handler resources that in turn has resources on ... ...

**Comment by khellang** [comment]
> So the Q is how to achieve this with MediatR in e.g. something like a long running windows service?

Just resolve the stuff you need from a lifetime scope (as shown in the blog post mentioned)? Make sure you dispose it when you're finished.

> As you would need to get hooked into before and after the func for resolving the handler....

I'm not sure what you're talking about here. What func?

> Just raising my voice as this easily could lead to people getting more and more handler resources that in turn has resources on ... ...

I don't think anyone here are experts on all the containers in the samples that have been contributed to this repo, and by all means, they could absolutely have bugs. I guess it's generally expected to know how the container you're using works. That should probably be a requirement anyway, using MediatR or not :wink: If you're having trouble wiring up the container properly, maybe it's best to just go "pure DI"-style? :smile:

**Comment by danielwertheim** [comment]
You want to be able to create the lifetime scope before the ....what you call "instance factories" are called and then be able to release the scope after a message has has been routed. Not the same lib, but you get the point: Something like this: http://danielwertheim.se/routemeister-and-autofac/

But MediatR probably has another lifecycle of the closure (call to the instance factories) giving you the actual instance of the handler. Or? Because I hope you aren't saying that if you want "per-request style resourced routing", you have to rebuild all the routes and the factory again for each request?

**Comment by khellang** [comment]
> You want to be able to create the lifetime scope before the ....what you call "instance factories" are called and then be able to release the scope after a message has has been routed.

Basically, you do what you want. MediatR doesn't dictate how you want to wire up your container, and I don't think it should. The sample is just that -- a sample, and it works fine.

> MediatR probably has another lifecycle of the closure (call to the instance factories) giving you the actual instance of the handler.

I'm not sure what you're talking about here. MediatR doesn't care about lifetime. It just calls one of the delegates passed to it when you send a request or notification, that's it! The rest is up to the user/container. It's open source, go look for yourself. The code base is a couple hundred lines of code, max :grin:

> Because I hope you aren't saying that if you want "per-request style resourced routing", you have to rebuild all the routes and the factory again for each request?

Again, I'm not sure what you mean with "routing" and "rebuild all the routes". Is this something in your application. It doesn't seem related to MediatR.

**Comment by danielwertheim** [comment]
Let me try again then... There seem to be no way to accomplish this, without recreating an instance of `Mediator`. Let me elaborate:

You inject the `InstanceFactories` "resolvers" in the constructor to `Mediator:
https://github.com/jbogard/MediatR/blob/master/src/MediatR/Mediator.cs#L29 and these resolvers would need access to the child scope, hence they need to be able to extract one tied to a certain message.

Suppose we do a `Publish` of two notifications: `Notification1` and `Notification2` using: https://github.com/jbogard/MediatR/blob/master/src/MediatR/Mediator.cs#L64

During this, an `InstanceFactory` is called to resolve the "handler" for each message. Now. To get the resolver to actually use a child scope, you would need to pass that e.g. as state with the message, so that the `InstanceFactory` could extract that and use it. Sure, you could potentially allow two hooks: `OnBeforeRoute` which could use `CallContext` and stash away arbitrary data (e.g. the scope); then extract that in the injected `InstanceFactory`, then release it in something like `OnAfterRoute`.

But given your answers there seem to be now way to accomplish this in MediatR. Hence you are either left to:

- Recreate `Mediator` for each message routed (guessing more penalty hits)
- Not expect to get a working solution for `Handler per message`, and use service locator pattern in your handlers.

And no, MediatR should not be tied to a IoC or something, but if there's no design decisions taken to provide tooling for the above scenario; you should clearly state: Either recreate `Mediator` instances or use some service locator inspired pattern in your handlers and keep them as singletons. So easy for people to think they are "safe" in setting the handlers as transient, but because how MediatR integrates with what ever you hook in, it can not be cleaned properly unless you use transient Mediator instances....

**Comment by jbogard** [comment]
What penalty hits for resolving per request?

**Comment by danielwertheim** [comment]
To clarify. This is not a bug and I'm not saying it will have a hit. It was a guess ("Guessing more penalty hits"). This cause you have handler caches etc in there that in that case would be set aside right? But I haven't looked closely. And yes, even if, it's most likely negligible, seen to all bits and pieces involved where MediatR would be used.

If this is the recommendation, to have the MediatR instance created on each request and thereby being able to manage the scope. I'm happy to have gotten this clarified for me.

Assuming usage of MediatR in conjunction with an IoC-container is a rather normal use-case, I still think you should emphasise this a bit more. Help people to use it right. But I can have missed that part trying it out.

**Comment by jbogard** [comment]
There's really not much of a recommendation here. I guess I do assume people use a container because otherwise it'd be a pain to build out the handler factories.

If the usage isn't clear, I'm happy to accept PRs for the readme or just edit the wikis directly. This is code I was copy-pasting amongst a dozen projects, so there might be a bit in my head about how I use it.

**Comment by leddt** [comment]
What if the Mediator's caches were changed to static fields? This way they would persist across instances, thus creating a new instance of the Mediator for each request would not cause any performance hit.

**Comment by khellang** [comment]
@leddt Yeah, I have no clue why they weren't static in the first place :smile:

**Comment by jbogard** [comment]
Just to spite you @khellang

**Comment by kappy** [comment]
Shouldn't it be the concern of how to configure the IMediator on the IoC (single instance vs instance per request) instead of without any usage choice, have the Dictionaries always on memory due their static nature?

**Comment by khellang** [comment]
@kappy The dictionaries are internal implementation details to cache types created using the reflection APIs. I don't think that's something the user needs to know or care about :smile:

**Comment by kappy** [comment]
@khellang Indeed, but then, on the other side, by making them static, you're limiting the their lifetime to be always in memory despite the lifetime configuration of the Mediator.

I didn't find any Thread-safety issues on the Mediator itself (I also didn't look too hard) so I don't see any issues on configure the Mediator as a Single Instance, something that Ioc containers already do

**Comment by leddt** [comment]
@kappy the problem with configuring it as a singleton is that the factory methods it receives as constructor parameter cannot use a per-request container, which is a very common pattern in web apps.

**Comment by leddt** [comment]
Also, note that the caches in question are simple type caches, not instance caches. Memory usage should be pretty small for that.

I guess we _could_ make the caches be another constructor dependency of the Mediator, which would enable users to configure the caches lifetime as they want, but then that would expose an implementation detail to the outside world, which is better kept hidden.

**Comment by kappy** [comment]
@leddt aren't the factory method just delegates? So all you're doing is querying the Ioc for the instances of the type, based on the scope it currently is.
Even if the Mediator is configured as Singleton, it could resolve the handler instances based on the current scope.

**Comment by leddt** [comment]
@kappy I'm only familiar with StructureMap, so maybe it's different with other containers. In the StructureMap sample for Mediator, the delegate's configuration looks like this:

```
cfg.For<SingleInstanceFactory>().Use<SingleInstanceFactory>(ctx => t => ctx.GetInstance(t));
cfg.For<MultiInstanceFactory>().Use<MultiInstanceFactory>(ctx => t => ctx.GetAllInstances(t));
```

If Mediator is a singleton, these delegates will be created only once, using the same container (the `ctx` variable in this code) as the Mediator itself.

But in StructureMap you create a _nested container_ to handle lifetime scopes. The nested container holds all of it's own instances, and refers to it's parent container to resolve longer lived objects (I'm probably explaining this terribly, I'm sorry).

So it is my understanding that, with StructureMap, using a singleton Mediator it would not be able to resolve request-scoped services. However it is possible that I am **completely** wrong about this. If that's the case then I guess the caches should indeed stay as instance fields.

**Comment by kappy** [comment]
@leddt You're not wrong. but then again, you can configure the delegates not form the context you're registering the Mediator, but from the actual Mvc (in this case) DependencyResolver. This way, the delegates will use the nested container that is created to handle the Request.

    For<SingleInstanceFactory>().Use<SingleInstanceFactory>(ctx => t => DependencyResolver.Current.GetService(t));
    For<MultiInstanceFactory>().Use<MultiInstanceFactory>(ctx => t => DependencyResolver.Current.GetServices(t));

I've written 2 samples, one using Autofac and another using StructureMap using exactly this configuration in [here](https://github.com/kappy/MediatRLifetimeTests).

**Comment by leddt** [comment]
@kappy I see! 

I wonder if this can work with Owin? The challenge with Owin is that there is not really a static way of getting the current request context. I'll have to think about it :smile:

**Comment by khellang** [comment]
> The challenge with Owin is that there is not really a static way of getting the current request context.

Some would argue that accessing a service locator in a static way like this is an anti-pattern :wink:

**Comment by kappy** [comment]
@leddt at least with Autofac, you can reach the scope with Owin using the `owinContext.GetAutofacLifetimeScope();` 

But I think we're getting a bit off topic. For me, the real question is if we sacrifice the Configuration easiness against keep a cache of types. 

Since this is not that a big size, and creating a Mediator instance per request is also cheap, I really don't see the point on making a hell out of configuration. 
The only disadvantage I see is if we change in runtime the container types registry (a bit like a plugin system for handlers). But this is clearly an edgy case.

I would say, go for this approach.
